File: v2/src/main/mongo/validator.ts
import type { Document, MongoClient } from 'mongodb';
import type { ValidatorPayload } from '../../shared/schema.js';

export async function getValidator(
  client: MongoClient,
  db: string,
  collection: string,
): Promise<ValidatorPayload> {
  const list = await client
    .db(db)
    .listCollections({ name: collection }, { nameOnly: false })
    .toArray();
  if (list.length === 0) return { validator: null };
  const opts = ((list[0] as Document)['options'] ?? {}) as Document;
  const v = opts['validator'];
  return {
    validator: v ? JSON.stringify(v, null, 2) : null,
    validationLevel: opts['validationLevel'] as ValidatorPayload['validationLevel'],
    validationAction: opts['validationAction'] as ValidatorPayload['validationAction'],
  };
}

export async function setValidator(
  client: MongoClient,
  db: string,
  collection: string,
  payload: ValidatorPayload,
): Promise<void> {
  const cmd: Document = { collMod: collection };
  if (payload.validator) {
    try {
      cmd['validator'] = JSON.parse(payload.validator);
    } catch (e) {
      throw new Error(`Validator is not valid JSON: ${(e as Error).message}`);
    }
  } else {
    cmd['validator'] = {};
  }
  if (payload.validationLevel) cmd['validationLevel'] = payload.validationLevel;
  if (payload.validationAction) cmd['validationAction'] = payload.validationAction;
  await client.db(db).command(cmd);
}

File: v3/src/main/kotlin/io/mex/mongo/Validator.kt
package io.mex.mongo

import com.mongodb.client.MongoClient
import org.bson.Document

data class ValidatorPayload(
    val validator: String?,
    val validationLevel: String?,
    val validationAction: String?,
)

fun getValidator(client: MongoClient, db: String, coll: String): ValidatorPayload {
    val infos = client.getDatabase(db).listCollections().filter(Document("name", coll)).toList()
    if (infos.isEmpty()) return ValidatorPayload(null, null, null)
    val opts = (infos[0]["options"] as? Document) ?: Document()
    val v = opts["validator"] as? Document
    return ValidatorPayload(
        validator = v?.toJson(),
        validationLevel = opts.getString("validationLevel"),
        validationAction = opts.getString("validationAction"),
    )
}

fun setValidator(client: MongoClient, db: String, coll: String, p: ValidatorPayload) {
    val cmd = Document("collMod", coll)
    cmd["validator"] = p.validator?.takeIf { it.isNotBlank() }?.let { Document.parse(it) } ?: Document()
    p.validationLevel?.let { cmd["validationLevel"] = it }
    p.validationAction?.let { cmd["validationAction"] = it }
    client.getDatabase(db).runCommand(cmd)
}
